New Resource: aws_api_gateway_documentation_part

radeksimko opened the pull request:
## Test results

```
$ make testacc TESTARGS='-run=TestAccAWSAPIGatewayDocumentationPart_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccAWSAPIGatewayDocumentationPart_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSAPIGatewayDocumentationPart_basic
--- PASS: TestAccAWSAPIGatewayDocumentationPart_basic (107.73s)
=== RUN   TestAccAWSAPIGatewayDocumentationPart_method
--- PASS: TestAccAWSAPIGatewayDocumentationPart_method (60.31s)
=== RUN   TestAccAWSAPIGatewayDocumentationPart_responseHeader
--- PASS: TestAccAWSAPIGatewayDocumentationPart_responseHeader (61.64s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	229.724s
```

bflad commented on aws/resource_aws_api_gateway_documentation_part.go:
```diff
@@ -0,0 +1,202 @@
+		RestApiId:           aws.String(d.Get("rest_api_id").(string)),
+	})
+	if err != nil {
+		if isAWSErr(err, "NotFoundException", "") {
```
Nitpick: Constant is available for this exception: `apigateway.ErrCodeNotFoundException`

bflad commented on aws/resource_aws_api_gateway_documentation_part_test.go:
```diff
@@ -0,0 +1,243 @@
+			resource.TestStep{
+				Config: testAccAWSAPIGatewayDocumentationPartConfig(apiName, properties),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSAPIGatewayDocumentationPartExists("aws_api_gateway_documentation_part.test", &conf),
```
Nitpick (personal): I've been finding it helpful to declare the resource names as a top level variable in the function so its easier to duplicate code across functions/resources. e.g. `resourceName := "aws_api_gateway_documentation_part.test"` - personal opinion though!

radeksimko replied:
I'm ok with that, sounds reasonable, the only thing is that the majority of the tests we have in the codebase now doesn't follow this, so we might want to revisit the existing code, so other folks copy-pasting code know what's "preferred".

bflad commented on aws/resource_aws_api_gateway_documentation_part_test.go:
```diff
@@ -0,0 +1,243 @@
+	var conf apigateway.DocumentationPart
+
+	rString := acctest.RandString(8)
+	apiName := fmt.Sprintf("tf_acc_api_%s", rString)
```
Generic question: should we try to include the test name (e.g. `t.Name()`) instead of the fairly diverse naming we have across resources for acceptance tests? I'm guessing it really depends on naming restrictions for some resources (max length, no underscores, etc).

radeksimko replied:
I agree we should make it test-specific, but `t.Name()` is a bit tricky for the reasons you mentioned, so I'll come up with custom, static name.

bflad commented on aws/resource_aws_api_gateway_documentation_part.go:
```diff
@@ -0,0 +1,202 @@
+		Delete: resourceAwsApiGatewayDocumentationPartDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
```
Must fix: missing acceptance test for import functionality

radeksimko replied:
Good catch! Will add.

bflad commented on aws/resource_aws_api_gateway_documentation_part.go:
```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		return err
+	}
+	d.SetId(*out.Id)
```
Does the documentation part ID need to be combined with the REST API ID to support import? e.g. `fmt.Sprintf("%s:%s", d.Get("rest_api_id").(string), *.out.Id)` and requisite `idParts := strings.Split(d.Id(), ":")`

radeksimko replied:
Yep, good catch, I'll pick `/` as the divider though. It seems safer, given that both IDs are then used in a URL, divided by that character.

bflad commented on website/docs/r/api_gateway_documentation_part.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+#### `location`
+
+* `method` - (Optional) The HTTP verb of a method. It is a valid field for the API entity types of `METHOD`, `PATH_PARAMETER`, `QUERY_PARAMETER`, `REQUEST_HEADER`, `REQUEST_BODY`, `RESPONSE`, `RESPONSE_HEADER`, and `RESPONSE_BODY`. The default value is `*` for any method.
```
Can we refer to the [API Gateway documentation](https://docs.aws.amazon.com/apigateway/api-reference/resource/documentation-part/) for this attribute and the ones below instead of listing these out?

bflad commented on website/docs/r/api_gateway_documentation_part.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+## Import
+
+API Gateway documentation_parts can be imported using the word `api-gateway-documentation_part`, e.g.
```
Needs to be updated for `(REST API ID and )documentation part ID` depending on outcome of ID value

bflad commented on aws/resource_aws_api_gateway_documentation_part_test.go:
```diff
@@ -0,0 +1,243 @@
+			return err
+		}
+
+		return fmt.Errorf("API Gateway Documentation Part %s/%s still exists.",
```
Should be updated based on ID discussion (or at least made a little clear what the second ID is). 😄

bflad commented on aws/resource_aws_api_gateway_documentation_part_test.go:
```diff
@@ -0,0 +1,243 @@
+		}
+		_, err := conn.GetDocumentationPart(req)
+		if err != nil {
+			if isAWSErr(err, "NotFoundException", "") {
```
Same note about `apigateway.ErrCodeNotFoundException` constant

bflad commented on aws/resource_aws_api_gateway_documentation_part_test.go:
```diff
@@ -0,0 +1,243 @@
+					testAccCheckAWSAPIGatewayDocumentationPartExists("aws_api_gateway_documentation_part.test", &conf),
+					resource.TestCheckResourceAttr("aws_api_gateway_documentation_part.test", "location.#", "1"),
+					resource.TestCheckResourceAttr("aws_api_gateway_documentation_part.test", "location.0.type", "API"),
+					resource.TestCheckResourceAttr("aws_api_gateway_documentation_part.test", "properties", `{"description":"Terraform Acceptance Test"}`),
```
Can this use `strconv.Unquote(properties)` instead?

radeksimko replied:
Good point in principle, I will use `strconv.Quote()`.

radeksimko commented:
@bflad I just addressed all of your comments. PTAL.

```
=== RUN   TestAccAWSAPIGatewayDocumentationPart_basic
--- PASS: TestAccAWSAPIGatewayDocumentationPart_basic (107.81s)
=== RUN   TestAccAWSAPIGatewayDocumentationPart_method
--- PASS: TestAccAWSAPIGatewayDocumentationPart_method (113.47s)
=== RUN   TestAccAWSAPIGatewayDocumentationPart_responseHeader
--- PASS: TestAccAWSAPIGatewayDocumentationPart_responseHeader (116.51s)
=== RUN   TestAccAWSAPIGatewayDocumentationPart_importBasic
--- PASS: TestAccAWSAPIGatewayDocumentationPart_importBasic (70.33s)
```

bflad commented:
This has been released in [terraform-provider-aws version 1.7.0](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#170-january-12-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
